crates/core_simd/src/array.rs

```diff
@@ -0,0 +1,165 @@
+use crate::intrinsics;
+use crate::masks::*;
+use crate::vector::ptr::SimdConst;
+use crate::vector::*;
+
+/// A representation of a vector as an "array" with indices,
+/// implementing operations applicable to any vector type and
+/// describing relationships between vector and scalar types.
+pub trait SimdArray<const LANES: usize>: crate::LanesAtMost32
+where
+    SimdUsize<LANES>: crate::LanesAtMost32,
+    SimdIsize<LANES>: crate::LanesAtMost32,
+    MaskSize<LANES>: crate::Mask,
+    Self: Sized,
+{
+    /// The scalar type in every lane of this vector type.
+    type Scalar: Copy + Sized;
+
+    /// Generates a SIMD vector with the same value in every lane.
+    #[must_use]
+    fn splat(val: Self::Scalar) -> Self;
+
+    /// SIMD gather: construct a SIMD vector by reading from a slice, using potentially discontiguous indices.
+    /// Out-of-bounds indices instead use the default value (zero, generally).
+    /// ```
+    /// # use core_simd::*;
+    /// let vec: Vec<i32> = vec![10, 11, 12, 13, 14, 15, 16, 17, 18];
+    /// let indices = SimdUsize::<4>::from_array([9, 3, 0, 5]);
+    ///
+    /// let result = SimdI32::<4>::gather(&vec, indices);
+    /// assert!(result == SimdI32::from_array([0, 13, 10, 15]));
+    /// ```
+    #[must_use]
+    #[inline]
+    fn gather(slice: &[Self::Scalar], idxs: SimdUsize<LANES>) -> Self
+    where
+        Self::Scalar: Default,
+    {
+        Self::gather_or(slice, idxs, Self::splat(Self::Scalar::default()))
+    }
+
+    /// SIMD gather: construct a SIMD vector by reading from a slice, using potentially discontiguous indices.
+    /// If an index is out of bounds, it instead selects from the values in the "or" vector.
+    /// ```
+    /// # use core_simd::*;
+    /// let vec: Vec<i32> = vec![10, 11, 12, 13, 14, 15, 16, 17, 18];
+    /// let indices = SimdUsize::<4>::from_array([9, 3, 0, 5]);
+    /// let alt = SimdI32::from_array([-5, -4, -3, -2]);
+    ///
+    /// let result = SimdI32::<4>::gather_or(&vec, indices, alt);
+    /// assert!(result == SimdI32::from_array([-5, 13, 10, 15]), "oh no, got {:?}", result);
+    /// ```
+    #[must_use]
+    #[inline]
+    fn gather_or(slice: &[Self::Scalar], idxs: SimdUsize<LANES>, or: Self) -> Self {
+        // IMPORTANT: We MUST construct our gather mask before we derive a pointer!
+        let mask = idxs.lanes_lt(SimdUsize::splat(slice.len())).to_int();
+        // IMPORTANT: We MUST obtain a pointer via as_ptr or an equivalent method!
+        let base_ptr = SimdConst::splat(slice.as_ptr());
```

Comment:
why is it important that these steps happen in this order?

Reply:
Oh, good point.

If you don't do it precisely in that order it makes the following calls potentially UB, I should make a note of that. :D

Reply:
ah, because len is &self, which invalidates the pointer if it's after?

incidentally, shouldn't this be as_mut_ptr?

Reply:
Yeah, re: &self.

Why would it be `as_mut_ptr`? This one is a gather, not a scatter. I am pretty sure obtaining `*mut T` from `&[T]` would be incorrect, unless we do in fact need to take `&mut [T]` to only do a gather. ^^;

Reply:
oh this is gather, right

Reply:
wait if this is gather and they're both &self methods then the ordering doesn't matter

Reply:
Oh OK~ Then I'll add those notes to the scatter. :D

Reply:
some more background: &mut T asserts it's unique, but &T does not, and so making more than one &T on the same value won't invalidate previous *const T you have to the same object.

This is not true for &mut T, with &mut T you *do* want to have the appropriate level of paranoia.

run it all through miri some day.